File: core/logger.py
"""
Structured Logging Module (Python)
Provides safe, structured logging without leaking credentials, passwords, or OTPs.
"""

import logging
import json
from pathlib import Path
from typing import Optional, Dict, Any, List
from core.security.sanitizer import redact_secrets_for_logs

LOGS_DIR = Path(__file__).resolve().parent.parent / "logs"
LOGS_DIR.mkdir(parents=True, exist_ok=True)
LOG_FILE = LOGS_DIR / "assistant.log"

# Circular buffer for UI endpoint viewing
_IN_MEMORY_LOGS: List[Dict[str, Any]] = []
MAX_IN_MEMORY_LOGS = 500


class SafeFormatter(logging.Formatter):
    def format(self, record: logging.LogRecord) -> str:
        orig = super().format(record)
        return redact_secrets_for_logs(orig)


def setup_logger(name: str = "assistant") -> logging.Logger:
    logger = logging.getLogger(name)
    if not logger.handlers:
        logger.setLevel(logging.DEBUG)

        handler = logging.FileHandler(str(LOG_FILE), encoding="utf-8")
        handler.setFormatter(
            SafeFormatter("[%(asctime)s] [%(levelname)s] [%(name)s] %(message)s")
        )
        logger.addHandler(handler)

        stream = logging.StreamHandler()
        stream.setFormatter(
            SafeFormatter("[%(asctime)s] [%(levelname)s] [%(name)s] %(message)s")
        )
        logger.addHandler(stream)
    return logger


assistant_logger = setup_logger()


def log_event(level: str, module: str, message: str, details: Optional[Dict[str, Any]] = None) -> None:
    clean_msg = redact_secrets_for_logs(message)
    entry = {
        "timestamp": logging.Formatter().formatTime(logging.LogRecord("", 0, "", 0, "", (), None)),
        "level": level.upper(),
        "module": module,
        "message": clean_msg,
        "details": details,
    }
    _IN_MEMORY_LOGS.insert(0, entry)
    if len(_IN_MEMORY_LOGS) > MAX_IN_MEMORY_LOGS:
        _IN_MEMORY_LOGS.pop()

    log_func = getattr(assistant_logger, level.lower(), assistant_logger.info)
    log_func(f"[{module}] {clean_msg}")


def get_logs(level: Optional[str] = None, limit: int = 100) -> List[Dict[str, Any]]:
    logs = _IN_MEMORY_LOGS
    if level:
        logs = [l for l in logs if l["level"] == level.upper()]
    return logs[:limit]

File: core/logger.ts
/**
 * Structured System Logger
 * Enforces privacy boundaries: NEVER logs API keys, passwords, OTPs, or private tokens.
 * Supports levels: DEBUG, INFO, WARN, ERROR.
 * Writes to memory buffer for UI viewing and appends to logs/assistant.log.
 */

import fs from 'node:fs';
import path from 'node:path';
import { redactSecretsForLogs } from './security/sanitizer.ts';

export type LogLevel = 'DEBUG' | 'INFO' | 'WARN' | 'ERROR';

export interface LogEntry {
  id: string;
  timestamp: string;
  level: LogLevel;
  module: string;
  message: string;
  details?: Record<string, any>;
}

class SystemLogger {
  private inMemoryLogs: LogEntry[] = [];
  private maxLogs: number = 500;
  private logFilePath: string | null = null;

  constructor() {
    try {
      if (
        typeof process !== 'undefined' &&
        process.cwd &&
        typeof path !== 'undefined' &&
        path?.resolve &&
        typeof fs !== 'undefined' &&
        fs?.existsSync
      ) {
        const logsDir = path.resolve(process.cwd(), 'logs');
        if (!fs.existsSync(logsDir)) {
          fs.mkdirSync(logsDir, { recursive: true });
        }
        this.logFilePath = path.join(logsDir, 'assistant.log');
      }
    } catch {
      this.logFilePath = null;
    }
  }

  private write(level: LogLevel, module: string, rawMessage: string, details?: Record<string, any>): LogEntry {
    const cleanMsg = redactSecretsForLogs(rawMessage);
    const sanitizedDetails = details
      ? JSON.parse(redactSecretsForLogs(JSON.stringify(details)))
      : undefined;

    const entry: LogEntry = {
      id: Math.random().toString(36).substring(2, 9),
      timestamp: new Date().toISOString(),
      level,
      module,
      message: cleanMsg,
      details: sanitizedDetails,
    };

    // Store in circular memory buffer
    this.inMemoryLogs.unshift(entry);
    if (this.inMemoryLogs.length > this.maxLogs) {
      this.inMemoryLogs.pop();
    }

    // Append to file asynchronously (safely ignoring errors in read-only setups)
    if (this.logFilePath && typeof fs !== 'undefined' && fs?.appendFileSync) {
      try {
        const line = `[${entry.timestamp}] [${level}] [${module}] ${cleanMsg} ${
          sanitizedDetails ? JSON.stringify(sanitizedDetails) : ''
        }\n`;
        fs.appendFileSync(this.logFilePath, line, 'utf8');
      } catch {
        // Ignore disk logging error gracefully
      }
    }

    // Console output for developer visibility
    const isTest = typeof process !== 'undefined' && process.env?.NODE_ENV === 'test';
    if (!isTest) {
      const prefix = `[${entry.timestamp}] [${level}] [${module}]`;
      if (level === 'ERROR') console.error(prefix, cleanMsg);
      else if (level === 'WARN') console.warn(prefix, cleanMsg);
      else console.log(prefix, cleanMsg);
    }

    return entry;
  }

  public debug(module: string, message: string, details?: Record<string, any>): LogEntry {
    return this.write('DEBUG', module, message, details);
  }

  public info(module: string, message: string, details?: Record<string, any>): LogEntry {
    return this.write('INFO', module, message, details);
  }

  public warn(module: string, message: string, details?: Record<string, any>): LogEntry {
    return this.write('WARN', module, message, details);
  }

  public error(module: string, message: string, details?: Record<string, any>): LogEntry {
    return this.write('ERROR', module, message, details);
  }

  public getLogs(level?: LogLevel, limit: number = 100): LogEntry[] {
    let result = this.inMemoryLogs;
    if (level) {
      result = result.filter((l) => l.level === level);
    }
    return result.slice(0, limit);
  }

  public clear(): void {
    this.inMemoryLogs = [];
  }
}

export const logger = new SystemLogger();
